src/core/server/elasticsearch/client/retry_call_cluster.ts

```diff
@@ -0,0 +1,113 @@
+import { concatMap, retryWhen } from 'rxjs/operators';
+import { errors as esErrors } from '@elastic/elasticsearch';
+import { ApiResponse, TransportRequestPromise } from '@elastic/elasticsearch/lib/Transport';
+import { Logger } from '../../logging';
+
+type ApiCaller = <TResponse, TContext>() => TransportRequestPromise<
+  ApiResponse<TResponse, TContext>
+>;
+
+const retryMigrationStatusCodes = [
+  503, // ServiceUnavailable
+  401, // AuthorizationException
+  403, // AuthenticationException
+  410, // Gone
+];
+
+/**
+ * Retries the provided Elasticsearch API call when a `NoLivingConnectionsError` error is
+ * encountered. The API call will be retried once a second, indefinitely, until
+ * a successful response or a different error is received.
+ *
+ * @example
+ * ```ts
+ * const response = await retryCallCluster(() => client.ping());
+ * ```
+ *
+ * @internal
+ */
+export const retryCallCluster = <TResponse, TContext>(
+  apiCaller: ApiCaller<TResponse, TContext>
+): TransportRequestPromise<ApiResponse<TResponse, TContext>> => {
+  return defer(() => apiCaller())
+    .pipe(
+      retryWhen((errors) =>
+        errors.pipe(
+          concatMap((error, i) =>
+            iif(
+              () => error instanceof esErrors.NoLivingConnectionsError,
+              timer(1000),
+              throwError(error)
+            )
+          )
+        )
+      )
+    )
+    .toPromise();
+};
+
+/**
+ * Retries the provided Elasticsearch API call when an error such as
+ * `AuthenticationException` `NoConnections`, `ConnectionFault`,
+ * `ServiceUnavailable` or `RequestTimeout` are encountered. The API call will
+ * be retried once a second, indefinitely, until a successful response or a
+ * different error is received.
+ *
+ * @example
+ * ```ts
+ * const response = await migrationRetryCallCluster(() => client.ping(), logger);
+ * ```
+ *
+ * @internal
+ */
+export const migrationRetryCallCluster = <TResponse, TContext>(
+  apiCaller: ApiCaller<TResponse, TContext>,
+  log: Logger,
+  delay: number = 2500
+): TransportRequestPromise<ApiResponse<TResponse, TContext>> => {
+  const previousErrors: string[] = [];
+  return defer(() => apiCaller())
+    .pipe(
+      retryWhen((errors) =>
+        errors.pipe(
+          concatMap((error, i) => {
+            if (!previousErrors.includes(error.message)) {
+              log.warn(`Unable to connect to Elasticsearch. Error: ${error.message}`);
+              previousErrors.push(error.message);
+            }
+            return iif(
+              () =>
+                error instanceof esErrors.NoLivingConnectionsError ||
+                error instanceof esErrors.ConnectionError ||
+                error instanceof esErrors.TimeoutError ||
+                retryMigrationStatusCodes.includes(error.statusCode) ||
+                error?.body?.error?.type === 'snapshot_in_progress_exception',
+              timer(delay),
+              throwError(error)
```

Comment:
This is a direct adaptation of https://github.com/elastic/kibana/blob/159369b71999b1f9d8df5645c0935044d25bb8bc/src/core/server/elasticsearch/legacy/retry_call_cluster.ts#L62-L74

@delvedor Could you take a look see if I did not do any mapping mistake?

Reply:
Looks good to me! :)

Reply:
The legacy client has a bug, where both the HTTP timeout error (408) and the socket timeout error will fall into the `RequestTimeout` error.
In the new client, the socket timeout is `TimeoutError`, while the HTTP timeout is a `ResponseError` with the `statusCode` property set to 408.
Not sure which one are you checking here.

Reply:
I'd say we probably want both. @rudolf can you confirm?

Reply:
Yes, we want to retry all operations, regardless of where the timeout occurred.

Reply:
From a performance point of view, instead of checking via `instanceof`, you could use `error.name === {name}`, which is an order of magnitude faster :)
Each error object of the client has a corresponding `.name` property: https://github.com/elastic/elasticsearch-js/blob/master/lib/errors.js